gemfire-core/src/main/java/com/gemstone/gemfire/cache/GemFireSparkConnectorCacheFactory.java

```diff
@@ -1,46 +1,157 @@
     super();
+    this.gfeGridMappings = null;
   }
 
-  public GemFireSparkConnectorCacheFactory(Properties props) {
+  public GemFireSparkConnectorCacheFactory(Properties props, Map<String, String> gfeGridMappings) {
     super(props);
+    this.gfeGridMappings = gfeGridMappings;
+
   }
 
 
-  public Cache create(PoolFactory pf)
-         throws TimeoutException, CacheWriterException,
+  public Cache create()
+      throws TimeoutException, CacheWriterException,
       GatewayException,
-      RegionExistsException
-  {
-       synchronized(CacheFactory.class) {
-          DistributedSystem ds = null;
-           if (this.dsProps.isEmpty()) {
-              // any ds will do
-                    ds = InternalDistributedSystem.getConnectedInstance();
-              }
-           if (ds == null) {
-               ds = DistributedSystem.connect(this.dsProps);
-              }
-            return GemFireSparkConnectorCacheImpl.create(pf, ds, cacheConfig);
+      RegionExistsException {
+    synchronized (CacheFactory.class) {
+      DistributedSystem ds = null;
+      if (this.dsProps.isEmpty()) {
+        // any ds will do
+        ds = InternalDistributedSystem.getConnectedInstance();
+      }
+      if (ds == null) {
+        ds = DistributedSystem.connect(this.dsProps);
+      }
+      PoolFactory defaultPF = null;
+       String defaultGrid =  gfeGridMappings.remove(GemFireSparkConnectorCacheImpl.gfeGridPropPrefix);
+      if (defaultGrid != null) {
+        defaultPF = this.createAndConfigurePoolFactory(defaultGrid);
+      }
+      Cache cache = GemFireSparkConnectorCacheImpl.create(defaultPF, gfeGridMappings, ds, cacheConfig);
+
+      for (Map.Entry<String, String> gridEntry : gfeGridMappings.entrySet()) {
+        String key = gridEntry.getKey().trim();
+        String gridName = key.substring(key.indexOf('.') + 1);
+        String locators = gridEntry.getValue().trim();
+        PoolFactory pf = this.createAndConfigurePoolFactory(locators);
+        pf.create(gridName);
+      }
+      return cache;
+    }
+  }
+
+
+  private PoolFactory createAndConfigurePoolFactory(String remoteLocators) {
+    PoolFactory pf = PoolManager.createFactory();
+    pf.setReadTimeout(30000);
+
+    StringTokenizer remoteLocatorsTokenizer = new StringTokenizer(remoteLocators, ",");
+    DistributionLocatorId[] locators = new DistributionLocatorId[remoteLocatorsTokenizer
+        .countTokens()];
+    int i = 0;
+    while (remoteLocatorsTokenizer.hasMoreTokens()) {
+      locators[i++] = new DistributionLocatorId(remoteLocatorsTokenizer.nextToken());
+    }
+    List<ServerLocation> servers = new ArrayList<ServerLocation>();
+    for (DistributionLocatorId locator : locators) {
+      try {
+
+        InetSocketAddress addr = new InetSocketAddress(locator.getHost(), locator.getPort());
+        GetAllServersRequest req = new GetAllServersRequest("");
+        Object res = TcpClient.requestToServer(addr.getAddress(), addr.getPort(), req, 2000);
```

Comment:
Why 2s timeout? Better to use something like DistributionConfig.getMemberTimeout()